Assignment/FSMTest.c

```diff
@@ -0,0 +1,152 @@
+// Training ~ A training program for new joiners at Metamation, Batch - July 2024.
+// Copyright (c) Metamation India.
+// Vinitha Perumal - GET
+// ------------------------------------------------------------------
+// Test.c
+// Program on A6B branch.
+// ------------------------------------------------------------------------------------------------
+#define _CRT_SECURE_NO_WARNINGS 1
+#include <malloc.h>
+#include <stdio.h>
+#include <string.h>
+#include <windows.h>
+#include <stdbool.h>
+
+int ExecProgram (char* exeFilePathAndName, const char* inputFilePathAndName, char* outputFilePathAndName);
+bool CompareFiles (char* file1, const char* file2, int* pos, char* ch1, char* ch2);
+
+/// <summary>
+/// This function will execute the FSM providing the input and output file names as arguments
+/// </summary>
+/// <param name="exeFilePathAndName"></param>
+/// <param name="inputFilePathAndName"></param>
+/// <param name="outputFilePathAndName"></param>
+/// <returns></returns>
+int ExecProgram (char* exeFilePathAndName, const char* inputFilePathAndName, char* outputFilePathAndName) {
+   char* cmdline = malloc (strlen (exeFilePathAndName) + strlen (inputFilePathAndName) + strlen (outputFilePathAndName) + 3);
+   if (cmdline == NULL) {
+      printf ("Unable to allocate memory\n");
+      return 1;
+   }
+   sprintf (cmdline, "%s%s%s%s%s", exeFilePathAndName, " ", inputFilePathAndName, " ", outputFilePathAndName);
+
+   // Set up structures for process information
+   STARTUPINFOA si;
+   PROCESS_INFORMATION pi;
+
+   // Initialize memory for STARTUPINFO and PROCESS_INFORMATION structures
+   ZeroMemory (&si, sizeof (si));
+   si.cb = sizeof (si);
+   ZeroMemory (&pi, sizeof (pi));
+
+   // Attempt to create the process
+   if (!CreateProcessA (
+      NULL,            // Application name (NULL if using command line)
+      cmdline,          // Command line (path to the executable)
+      NULL,            // Process handle not inheritable
+      NULL,            // Thread handle not inheritable
+      FALSE,           // Set handle inheritance to FALSE
+      0,               // No creation flags
+      NULL,            // Use parent's environment block
+      NULL,            // Use parent's starting directory
+      &si,             // Pointer to STARTUPINFO structure
+      &pi))            // Pointer to PROCESS_INFORMATION structure
+   {
+      // If CreateProcess fails, print an error message
+      int err = GetLastError ();
+      printf ("Failed to start process. Error: %lu\n", GetLastError ());
+      free (cmdline);
+      return 1;
+   }
+
+   // Wait until the process has finished execution
+   WaitForSingleObject (pi.hProcess, INFINITE);
+
+   // Close handles
+   CloseHandle (pi.hProcess);
+   CloseHandle (pi.hThread);
+
+   free (cmdline);
+   return 0;
+
+}
+
+bool CompareFiles (char* file1, const char* file2, int* pos, char* ch1, char* ch2) {
+   FILE* f1 = fopen (file1, "r"), * f2 = fopen (file2, "r");

```

Comment:
FYI: Since we know for a fact this is a file, we are reading the entire file at once. Had that not been the case (like receiving bits over the net) we'd be reading char by char (since we may not know how many bits we'd be reading.